src/main/java/org/jenkinsci/plugins/vsphere/FixedLifespanCloudRetentionStrategy.java

```diff
@@ -0,0 +1,114 @@
+import hudson.slaves.RetentionStrategy;
+import hudson.util.TimeUnit2;
+
+import java.io.IOException;
+import java.util.concurrent.ExecutionException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+import static java.util.logging.Level.WARNING;
+
+public final class FixedLifespanCloudRetentionStrategy extends RetentionStrategy<AbstractCloudComputer> {
+
+    private static final Logger LOGGER = Logger.getLogger(CloudRetentionStrategy.class.getName());
+
+    private final int lifespanMinutes;
+
+    @DataBoundConstructor
+    public FixedLifespanCloudRetentionStrategy(int lifespanMinutes) {
+        this.lifespanMinutes = lifespanMinutes;
+    }
+
+    public int getLifespanMinutes() {
+        return lifespanMinutes;
+    }
+
+    @Override
+    public long check(AbstractCloudComputer c) {
+        final long creationTimeMillis = c.getAction(CloudComputerCreatedOnInvisibleAction.class).getCreationTimeMillis();
+        final long ageMillis = System.currentTimeMillis() - creationTimeMillis;
+        final String cname = c.getName();
+        if (!isAtEndOfLife() && ageMillis > TimeUnit2.MINUTES.toMillis(lifespanMinutes)) {
+            LOGGER.log(Level.FINE, "Will terminate {0} once idle - lifespan of {1} minutes reached.", new Object[] { cname, lifespanMinutes });
+            setAtEndOfLife();
+            final VSphereOfflineCause cause = new VSphereOfflineCause(Messages._fixedLifespanCloudRetentionStrategy_OfflineReason_LifespanReached(String.valueOf(lifespanMinutes)));
+            try {
+                c.disconnect(cause).get();
+            } catch (InterruptedException | ExecutionException e) {
+                LOGGER.log(WARNING, "Failed to disconnect " + cname, e);
+            }
+        }
+        if (isAtEndOfLife() && c.isIdle()) {
+            final AbstractCloudSlave computerNode = c.getNode();
+            if (computerNode != null) {
+                try {
+                    LOGGER.log(Level.FINER, "Initiating termination of {0}.", cname);
+                    computerNode.terminate();
+                } catch (InterruptedException | IOException e) {
+                    LOGGER.log(WARNING, "Failed to terminate " + cname, e);
+                }
+            }
+        }
+        return 1; // re-check in 1 minute
+    }
+
+    @Override
+    public DescriptorImpl getDescriptor() {
+        return DESCRIPTOR;
+    }
+
+    @Override
+    public void start(AbstractCloudComputer c) {
+        c.addAction(new CloudComputerCreatedOnInvisibleAction(System.currentTimeMillis()));
+        super.start(c);
+        c.connect(false);
+    }
+
+    @Override
+    public boolean isAcceptingTasks(AbstractCloudComputer c) {
+        return !isAtEndOfLife();
+    }
+
+    private transient boolean atEndOfLife;
+
+    private synchronized boolean isAtEndOfLife() {
```

Comment:
I don't think that `synchronized` _here_ is going to achieve thread-safety.
e.g. the `check` method calls `isAtEndOfLife()` twice, so if you only lock in here, `check` might get two different values when it calls the method if another thread has called `setAtEndOfLife()` in between.
You probably want to make `check()` be `synchronized` (or add a `synchronized` section within it) so that you can ensure that you get a consistent and controlled logic flow within that method.
...also, if you put it there, you can reduce the number of synchronized blocks to just one - thread synchronization is expensive so we don't want to do that any more than we absolutely have to.
...and I'd suggest that you check with other retention strategies (in this and other plugins) and verify that it's needed at all because, if we don't need to make this code thread-safe, we shouldn't add code making it look like it is.